src/qibolab/instruments/zhinst.py

```diff
@@ -294,6 +321,13 @@ def __init__(self, name, descriptor, use_emulation=False):
         "Storing sweepers"
         # Improve the storing of multiple sweeps
 
+        self.ports = {}
+
+    def __getitem__(self, port_name: Tuple[str, str]):
+        if port_name not in self.ports:
+            self.ports[port_name] = ZhPort(*port_name)
+        return self.ports[port_name]
```

Comment:
These lines are repeated essentially the same for every controller, the main differences are:

1. the port class
2. how arguments are passed to the port class
3. the type hints

About 1. and 2. this would be easily solvable: instead of defining the whole `_getitem`, let's just make a `staticmethod` that given the `port_name` will return the port object.
For the RFSoC this could just be assigned to the port class itself, since it's already respecting the signature, while here for Zurich it's just a matter of making a tiny wrapper to add the `*`.

3. could be solved with generics, but, in a first iteration, we could even forget about it.

Reply:
I was also thinking to move this to the abstract `Instrument` and not have it defined seperately in each driver. There are even workarounds to remove the `*` and make it equivalent in all drivers, so this is certainly possible to do if we forget about (3).

I only didn't do it for psychological reasons. Now I force driver developers to code this themeselves and understand the mechanism of getting ports from the instrument. If it is inherited automatically, it may not be clear to everyone what is happening. But since you agree, I will lift it to the base class.